### DISCUSSION TITLE
Tesla API -> Oauth2 and MFA

### DISCUSSION THREAD

**Discussion by LelandSindt** [teslamate-org/teslamate#981] [General]
@adriankumpf have you been following this? 

https://github.com/timdorr/tesla-api/issues/215

**Comment by adriankumpf** [comment]
Yup, I have been observing this for a while. I am prepared to make changes if necessary. But so far it looks like there is nothing to do yet.

**Comment by natrlhy** [comment]
I haven't enabled MFA on my tesla account yet, but I also don't want it to break TeslaMate if I do :)

**Comment by cwanja** [comment]
> Yup, I have been observing this for a while. I am prepared to make changes if necessary. But so far it looks like there is nothing to do yet.

I will jump in and enable it. See what all I can break 😊

**Comment by andreask** [comment]
> I will jump in and enable it. See what all I can break

Have you seen anything yet? I really want to add MFA to my Tesla account.

**Comment by cwanja** [comment]
> > I will jump in and enable it. See what all I can break
> 
> Have you seen anything yet? I really want to add MFA to my Tesla account.

Two days in @andreask and no issues. TeslaMate has lost no connection to my car. Per the original link, does not look like Tesla deprecated the old authentication mechanisms, but probably will at some point. At that point, I believe there will need to be updates.

**Comment by andreask** [comment]
> > > I will jump in and enable it. See what all I can break
> > 
> > 
> > Have you seen anything yet? I really want to add MFA to my Tesla account.
> 
> Two days in @andreask and no issues. TeslaMate has lost no connection to my car. Per the original link, does not look like Tesla deprecated the old authentication mechanisms, but probably will at some point. At that point, I believe there will need to be updates.

@cwanja 
The thing is I worry about logging in "again" - or if the refresh token expires?

**Comment by LexNastin** [comment]
> > > > I will jump in and enable it. See what all I can break
> > > 
> > > 
> > > Have you seen anything yet? I really want to add MFA to my Tesla account.
> > 
> > 
> > Two days in @andreask and no issues. TeslaMate has lost no connection to my car. Per the original link, does not look like Tesla deprecated the old authentication mechanisms, but probably will at some point. At that point, I believe there will need to be updates.
> 
> @cwanja
> The thing is I worry about logging in "again" - or if the refresh token expires?

I was messing around with the api... and it seems like at the moment, nothing in the api requires 2fa/mfa... i even tried remote starting the car and removing pin to drive... nothing ever asked for 2fa/mfa. So at the moment, it doesn't secure anything except the main app and the website, so right now... there's no need to worry TeslaMate and the api... when tesla fully implement it, i will mess around with the car again and see if anything changed.

**Comment by FlyingLemming** [comment]
I am afraid I have to disagree. ;-)
Indeed, the login – or rather the token retrieval – still works with the old procedure. However, only one app can be used at a time. As soon as any other app or service gets a token for itself, the previous token is invalidated right away.
Some services (TeslaFi, Tronity, Watch for Tesla) have updated to the new MFA procedure (which can also be used by users not having MFA activated yet) and they run unaffected by other services. However, they do disconnect other services using the legacy login method. ABRP has updated as well and gets along with the beforementioned services, but it is still getting disconnected whenever a legacy app retrieves a token. 

Hence, old-schoold apps or services only work correctly when they are the only service in use.
This may not apply to all users yet. An e-mail I received from TeslaFi regarding this issue suggests that Tesla is only updating the user accounts gradually. My account has been affected for about six weeks now although I do not have MFA activated yet.

tl;dr: It may be advisable to update all apps and services to the new MFA login or token retrieval procedure asap to make sure they are not affected by and do not affect other apps using the Tesla API.

**Comment by okamiokami** [comment]
@adriankumpf any updates (march 24th, 2021)? Thank you

**Comment by Hxd097** [comment]
Any update on MFA support？

**Comment by okamiokami** [comment]
Teslamate worked for me even with MFA active.

I have been running it for a few months, no need to reauthenticate or anything.

No other apps to make the token expire, so my scenario might not be very representative.

**Comment by wvalko** [comment]
I'm currently getting a "Error: Captcha does not match" when trying to login to Teslamate.  I have 2FA enabled on my account.  seems like there is no way to disable it either.  any solutions here?

**Comment by ardadonmezer** [comment]
I'm getting "Error: :mfa_input_not_found," tried both with MFA enabled and disabled.
It was a clean install, and no other apps connected before with this account except the tesla app.

**Reply by bluey47** [reply]
I am also getting "Error: :mfa_input_not_found". 
I am a first time user (picked up car last week), really wanted to just get this working after spending days getting docker working (long not relevant story).
I have bypassed the login by using an android App called Tesla Tokens to generate the token and this allowed Teslamate to work. I don't like that solution, so will change my Tesla password once normal token generation is fixed.

**Reply by casmith** [reply]
Same issue even after disabling MFA on my Tesla account. I have no android device to work around this with.

**Reply by bluey47** [reply]
I am sure there is an equivalent IOS app out there. It is worth it, Teslamate is amazing, I am loving all the stats.

**Reply by bluey47** [reply]
Token apps are listed here as well as the current issue. https://github.com/adriankumpf/teslamate/issues/1945#issuecomment-913684723

**Comment by Gunthermic** [comment]
So I take it Tesla mate is dead in water... without tryin to use a tolken from one of those other apps.

**Reply by bluey47** [reply]
See https://github.com/adriankumpf/teslamate/issues/1945#issuecomment-913684723
